File: ruby/reverse_string.rb
def reverse_string(str)
  reversed = ""
  str.chars.each do |char|
    reversed = char + reversed
  end

  reversed
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'ih'"
  puts "=>", reverse_string('hi')

  puts

  puts "Expecting: 'ybabtac'"
  puts "=>", reverse_string('catbaby')

  puts "Expecting: '!dlroW ,olleH'"
  puts "=>", reverse_string("Hello, World!")
end

# Iterate through string, and add each new character to new string
# plus the remainder of the reversed string... This way the reversed string 
# gets the new chars added to the beginning, effectively reversing the order.


# The reversed string starts as an empty string "", and on the first iteration,
# the first char is added. Now the string has one char, the first char in the
# original string. On subsequent iterations, the reversed string is added to
# each new char in original order, thus pushing the first chars to the end and 
# reversing the order of the string.
# EX: "Hello" => "H" + ""; "ello" => "e" + "H"; "llo" => "l" + "eH" ... "olleH"

File: javascript/reverse_string.js
function reverseString(str) {
  let reversed = "";
  for (let i = str.length - 1; i > -1; --i) {
    reversed = reversed + str[i];
  }
  return reversed;
}

console.log(reverseString("Hello, world!"))

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'ih'");
  console.log("=>", reverseString("hi"));

  console.log("");

  console.log("Expecting: 'ybabtac'");
  console.log("=>", reverseString("catbaby"));
}

module.exports = reverseString;

// iterate backwards through string (array), then set reversed to reversed + current
// return reversed


// Iterating over the string (array) backwards will give the values according
// to the negative index. Those can then be added to the array in reverse order
// so when the value is returned, it should be in reverse order.
